### PR TITLE
Support custom main file .scry_main.cr and add some diagnostic fixes

### PR DESCRIPTION
**Pull request by faustinoaq**
This PR depends on #94 (Should be merged only after that)

Related to #80

Currently to support diagnostics with implicit requires, scry is already using a "ghost" file `.scry_main.cr`,  to analyze files inside `./src/*`. although, this use case can change, and people that don't use src folder still can use implicit requires by adding a `.scry_main.cr` file with their `require`s on workspace root

screenshot_20180423_142417

### CODE REVIEW

**Comment by faustinoaq** [comment]
This PR should be updated with master after #94 is merged

File: src/scry/publish_diagnostic.cr [hunk 1]
```diff
@@ -6,7 +6,8 @@ require "./workspace"
 
 module Scry
   struct PublishDiagnostic
-    METHOD = "textDocument/publishDiagnostics"
+    METHOD                = "textDocument/publishDiagnostics"
+    FILES_WITH_DIAGNOSTIC = [] of String
```
Comment (bew) [inline src/scry/publish_diagnostic.cr]: Why it that a constant and not an ivar?

Comment (faustinoaq) [reply src/scry/publish_diagnostic.cr]: @bew Oh, I just realize we don't need that, we don't need even a var, this is used just on one place, let me fix that :+1:

Comment (faustinoaq) [reply src/scry/publish_diagnostic.cr]: Done!

Comment (bew) [reply src/scry/publish_diagnostic.cr]: Huh, I was talking about `FILES_WITH_DIAGNOSTIC` :joy:

File: src/scry/publish_diagnostic.cr [hunk 2]
```diff
@@ -6,20 +6,31 @@ require "./workspace"
 
 module Scry
   struct PublishDiagnostic
-    METHOD = "textDocument/publishDiagnostics"
+    FILES_WITH_DIAGNOSTIC = [] of String
```
Comment (bew) [inline src/scry/publish_diagnostic.cr]: huh I was talking about this constant :joy:

Comment (faustinoaq) [reply src/scry/publish_diagnostic.cr]: Well, this constant is required, `FILES_WITH_DIAGNOSTIC` is updated only during `@diagnostic.from`, so, if I use an ivar, then next time I fix my code, I won't be able to clean the past errors :sweat_smile: 

I could use a class var as well, but I think a constant is cleaner

File: src/scry/publish_diagnostic.cr (hunk 1 above)
Comment (faustinoaq) [reply src/scry/publish_diagnostic.cr]: See https://github.com/crystal-lang-tools/scry/pull/95#discussion_r183521067

File: src/scry/publish_diagnostic.cr (hunk 2 above)
Comment (bew) [reply src/scry/publish_diagnostic.cr]: Oh ok, could you rename it as `FILES_WITH_DIAGNOSTIC` with a little comment about that?

Comment (faustinoaq) [reply src/scry/publish_diagnostic.cr]: @bew Done!, I just discovered this was clearing all diagnostics, even those not generated by a current file, so, I'm using a Hash now to group diagnostic per file:

Imgur

**Comment by faustinoaq** [comment]
#94 is already merged! :tada: waiting for more reviews :+1:

**Comment by faustinoaq** [comment]
@bew What about just `.main.cr` ?

**Comment by bew** [comment]
No, I prefer to have `scry` in the name, so I know why this file is there

**Pull request merged**